### PR TITLE
DiskRP Swagger changes for 2020-12-01

### PR DESCRIPTION
**Pull request by meet1993shah**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [x] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

SDK PR - https://github.com/Azure/azure-sdk-for-net/pull/19092

### CODE REVIEW

File: custom-words.txt [hunk 1]
```diff
@@ -131,6 +131,7 @@ artifacturi
 asciifolding
 Asns
 aspnet
+ASSDZRS
```
Comment (chetmaddula) [inline custom-words.txt:134]: why is this needed?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json [hunk 2]
```diff
@@ -2026,7 +2067,7 @@
           }
         },
         "purchasePlan": {
-          "$ref": "#/definitions/PurchasePlan",
+          "$ref": "#/definitions/DiskPurchasePlan",
```
Comment (chetmaddula) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:2070]: why the name change? should we have a SnapshotPurchasePlan?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/GetInformationAboutAManagedDisk.json [hunk 3]
```diff
@@ -20,6 +20,7 @@
             "publisher": "test_test_pmc2pc1",
             "product": "marketplace_vm_test"
           },
+          "supportsHibernation": true,
```
Comment (chetmaddula) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/GetInformationAboutAManagedDisk.json:23]: should securityProfile be added here?

File: custom-words.txt (hunk 1 above)
Comment (meet1993shah) [reply custom-words.txt:134]: SDK validation threw error for the new SSD ZRS test added, so the options were to change the test name or add the custom word.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json (hunk 2 above)
Comment (meet1993shah) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:2070]: PurchasePlan definition already exists for other compute resources, we just needed a different defnition name, thus DiskPurchasePlan.  PurchasePlan for snapshot uses the same defintion so no need to add SnapshotPurchasePlan.

File: custom-words.txt (hunk 1 above)
Comment (chetmaddula) [reply custom-words.txt:134]: ASSDZRS isnt custom word.. either add it as SSDZRS or change the test name to CreateAManagedDiskWithSSDZRSAccountType.json

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/GetInformationAboutAManagedDisk.json (hunk 3 above)
Comment (meet1993shah) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/GetInformationAboutAManagedDisk.json:23]: @anshulsolanki21

Comment (anshulsolanki21) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/GetInformationAboutAManagedDisk.json:23]: Raised PR for this change : https://github.com/meet1993shah/azure-rest-api-specs/pull/14

File: package-lock.json [hunk 4]
```diff
@@ -2718,8 +2718,7 @@
         "esprima": "^4.0.1",
         "estraverse": "^4.2.0",
         "esutils": "^2.0.2",
-        "optionator": "^0.8.1",
-        "source-map": "~0.6.1"
+        "optionator": "^0.8.1"
```
Comment (chetmaddula) [inline package-lock.json:2721]: why are these needed?

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json [hunk 5]
```diff
@@ -0,0 +1,3432 @@
...
+        "tier": {
+          "type": "string",
+          "description": "Performance tier of the disk (e.g, P4, S10) as described here: https://azure.microsoft.com/en-us/pricing/details/managed-disks/. Does not apply to Ultra disks."
+        },
+        "burstingEnabled": {
+          "type": "boolean",
+          "description": "Set to true to enable bursting beyond the provisioned performance target of the disk. Bursting is disabled by default. Does not apply to Ultra disks."
+        },
+        "purchasePlan": {
+          "$ref": "#/definitions/DiskPurchasePlan",
```
Comment (ArcturusZhang) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:2425]: We are changing this type from PurchasePlan to DiskPurchasePlan because they remove the PromotionCode in the original `PurchasePlan` model and we need it here?
I am asking this because this is causing a SDK breaking change across the api-versions

Comment (meet1993shah) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:2425]: the DiskRP backend still supports PromotionCode, so this will be needed. @vikramd-ms what should we do here?

Comment (meet1993shah) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:2425]: also, why is this causing breaking change? the definition did not change, only the name of the definition was changed here.

Comment (Sandido) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:2425]: Changing the type name changes method signatures and makes the code think the type changed, making it a breaking change.
Please revert this change.

Comment (vikramd-ms) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:2425]: We will need to use DiskPurcahsePlan because it contains PromotionCode while PurchasePlan does not and DiskRP does return a PromotionCode property. So we are trying to *avoid* a breaking change here. @ArcturusZhang, please clarify how using DiskPurchasePlan is causing a breaking change and how we should proceed.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json [hunk 6]
```diff
@@ -0,0 +1,3432 @@
...
+      "description": "disk encryption set resource."
+    },
+    "DiskEncryptionSetUpdate": {
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/DiskEncryptionSetUpdateProperties"
+        },
+        "identity": {
+          "$ref": "#/definitions/EncryptionSetIdentity"
```
Comment (Sandido) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:2857]: when adding a new optional property, please add it to the end of the property list.

File: specification/compute/resource-manager/readme.md [hunk 7]
```diff
@@ -218,7 +218,7 @@ input-file:
 - Microsoft.Compute/stable/2020-12-01/compute.json
 - Microsoft.Compute/stable/2020-12-01/runCommands.json
 - Microsoft.Compute/stable/2019-04-01/skus.json
-- Microsoft.Compute/stable/2020-09-30/disk.json
+- Microsoft.Compute/stable/2020-12-01/disk.json
```
Comment (Sandido) [inline specification/compute/resource-manager/readme.md:221]: Please generate the .Net SDK code and create tests for the new functionality, record those tests, and include the recorded json files in the PR. 
Please link to the .Net SDK PR here. The tests must all pass before this Swagger PR can be merged.

**Review by Sandido — CHANGES_REQUESTED** [review]

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json (hunk 5 above)
Comment (Sandido) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:2425]: The swagger definition of [ PurchasePlan has PromotionCode](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/compute/resource-manager/Microsoft.Compute/stable/2020-09-30/disk.json#L2588), is that incorrect? 
Why is the name change needed?

Comment (Sandido) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:2425]: Oh, are you referring to the compute.json definition?
https://github.com/Azure/azure-rest-api-specs/blob/master/specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json#L8340

Comment (chetmaddula) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:2425]: since we're not using that definition, it should be fine to use PurchasePlan itself right?

Comment (laurawu19) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json]: Use the PurchasePlan in disk.json will trigger the conflicts with compute.json because different definition (with/without promotion code). DiskPurchasePlan is added to define the purchase plan on diskRP side. It is not a name for parameter, and it is a name for defination. We have done that for diskEncryptionSet before.

File: specification/compute/resource-manager/readme.md (hunk 7 above)
Comment (meet1993shah) [reply specification/compute/resource-manager/readme.md:221]: https://github.com/Azure/azure-sdk-for-net/pull/19092

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json (hunk 5 above)
Comment (meet1993shah) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json]: I reverted the change back to PurchasePlan, and also ran the tests again, all the tests in sdk pass. The only check failing is the SDK track 2 validation, and since its not a required check this change should be fine. 
@ArcturusZhang , @Sandido. Please review this PR again. Have also attached the dot net sdk PR - https://github.com/Azure/azure-sdk-for-net/pull/19092

**Comment by ArcturusZhang** [comment]
We are having `PurchasePlan` in disk.json and compute.json, but these two definitions are not identical. 
Is there a way that we could work this around without introducing breaking changes? @Sandido any thoughts?

**Review by ArcturusZhang — APPROVED** [review]
Approved since the Track 2 Validation error exists before this PR

File: specification/compute/resource-manager/readme.md (hunk 7 above)
Comment (Sandido) [reply specification/compute/resource-manager/readme.md:221]: @meet1993shah , have you investigated the .Net SDK failures? Seems like the build is failing.

Comment (meet1993shah) [reply specification/compute/resource-manager/readme.md:221]: resolved the errors, build is passing

**Review by Sandido — APPROVED** [review]

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateADiskEncryptionSet.json [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+          },
+          "keyUrl": "https://myvmvault.vault-int.azure-int.net/keys/{key}"
+        },
+        "encryptionType": "EncryptionAtRestWithCustomerKey"
+      }
+    }
+  },
+  "responses": {
+    "202": {
+      "body": {
```
Comment (j5lim) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateADiskEncryptionSet.json:25]: >"body": { [](start = 6, length = 9)

RP should always return top level properties like Type and Id.

**Comment by j5lim** [comment]
>     },

I would recommend referencing the common type json.

---
Refers to: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:1728 in beef77b. [](commit_id = beef77bc1167e286f9f104d85e4a97388dda9e70, deletion_comment = False)

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/UpdateAManagedDiskToAddSupportsHibernation.json [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  "responses": {
+    "202": {
+      "body": {
+        "name": "myDisk",
+        "location": "westus",
+        "sku": {
+          "name": "Standard_LRS",
+          "tier": "Standard"
+        },
+        "properties": {
```
Comment (j5lim) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/UpdateAManagedDiskToAddSupportsHibernation.json:22]: Does PATCH api return tags as well? It should if not.

**Comment by j5lim** [comment]
>             },

Please use the same schema for PUT and PATCH. This is critical for platform features like Azure policy, ARG, etc.
[https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc008-put-get-patch--list-must-return-the-same-resource-schema](https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc008-put-get-patch--list-must-return-the-same-resource-schema)

You can use x-ms-mutability or read-only to mark properties for creation-only or read-only.
[https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi028-annotate-immutable-properties-with-x-ms-mutability](https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi028-annotate-immutable-properties-with-x-ms-mutability)
[https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi019-annotate-read-only-properties-with-readonly-true](https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi019-annotate-read-only-properties-with-readonly-true)

---
Refers to: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:149 in beef77b. [](commit_id = beef77bc1167e286f9f104d85e4a97388dda9e70, deletion_comment = False)

**Comment by j5lim** [comment]
>               "$ref": "#/definitions/SnapshotUpdate"

Please use the same schema for PUT and PATCH. (please apply it to all other resource types)

---
Refers to: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:489 in beef77b. [](commit_id = beef77bc1167e286f9f104d85e4a97388dda9e70, deletion_comment = False)

**Comment by j5lim** [comment]
>           "202": {

201 + provisingState + Azure-AsyncOperation header is the recommended pattern for PUT.
[https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc0025-201-is-the-recommended-async-pattern](https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc0025-201-is-the-recommended-async-pattern)


---
Refers to: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:1099 in beef77b. [](commit_id = beef77bc1167e286f9f104d85e4a97388dda9e70, deletion_comment = False)

**Comment by j5lim** [comment]
>               "$ref": "#/definitions/DiskAccess"

202 should return headers only.

---
Refers to: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:1102 in beef77b. [](commit_id = beef77bc1167e286f9f104d85e4a97388dda9e70, deletion_comment = False)

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/ApprovePrivateEndpointConnection.json [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+      "properties": {
+        "privateLinkServiceConnectionState": {
+          "status": "Approved",
+          "description": "Approving myPrivateEndpointConnection"
+        }
+      }
+    }
+  },
+  "responses": {
+    "202": {
```
Comment (j5lim) [inline specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/ApprovePrivateEndpointConnection.json:18]: >   "202" [](start = 1, length = 8)

It should return 201 if the initial async call returns the resource with provisioning state. Please follow the RPC.
[https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creatingupdating-using-put](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creatingupdating-using-put)

**Comment by j5lim** [comment]
>     "CloudError": {

I recommend referencing the common type json.

---
Refers to: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:3252 in beef77b. [](commit_id = beef77bc1167e286f9f104d85e4a97388dda9e70, deletion_comment = False)

**Review by j5lim — CHANGES_REQUESTED** [review]
left comments.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/ApprovePrivateEndpointConnection.json (hunk 10 above)
Comment (meet1993shah) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/ApprovePrivateEndpointConnection.json:18]: @vikramd-ms

Comment (coscor-ms) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/ApprovePrivateEndpointConnection.json:18]: The requirement to return 201 from PUT is a recent change to the protocol (it was checked in 17 months ago); the original RPC allowed either 201 or 202 to be returned at resource creation (this can still be seen parts of the addendum which should be clarified).
Changing from 202 to 201 requires an API version change; we'll take it as a work item, but I think it's outside the scope of this checkin

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/UpdateAManagedDiskToAddSupportsHibernation.json (hunk 9 above)
Comment (meet1993shah) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/UpdateAManagedDiskToAddSupportsHibernation.json:22]: PATCH api does return tags. However the examples in question dont have tags in the request or the resource, so tags are not present in the response object.

**Comment by meet1993shah** [comment]
> > ```
> >         },
> > ```
> 
> Please use the same schema for PUT and PATCH. This is critical for platform features like Azure policy, ARG, etc.
> https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc008-put-get-patch--list-must-return-the-same-resource-schema
> 
> You can use x-ms-mutability or read-only to mark properties for creation-only or read-only.
> https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi028-annotate-immutable-properties-with-x-ms-mutability
> https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi019-annotate-read-only-properties-with-readonly-true
> 
> Refers to: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:149 in [beef77b](https://github.com/Azure/azure-rest-api-specs/commit/beef77bc1167e286f9f104d85e4a97388dda9e70). [](commit_id = [beef77b](https://github.com/Azure/azure-rest-api-specs/commit/beef77bc1167e286f9f104d85e4a97388dda9e70), deletion_comment = False)

The schema for PUT and PATCH are the same, even though there are two separate names, this has been followed on from the earlier version, its not a new change. However, if we try to use the same name schema now, it might cause a breaking change which we want to avoid. The current schema for PUT and PATCH are the same and wont cause any break in policy or ARG.

**Comment by meet1993shah** [comment]
> > ```
> >           "$ref": "#/definitions/DiskAccess"
> > ```
> 
> 202 should return headers only.
> 
> Refers to: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:1102 in [beef77b](https://github.com/Azure/azure-rest-api-specs/commit/beef77bc1167e286f9f104d85e4a97388dda9e70). [](commit_id = [beef77b](https://github.com/Azure/azure-rest-api-specs/commit/beef77bc1167e286f9f104d85e4a97388dda9e70), deletion_comment = False)

I think based on the earlier versions we always return the schema with the headers for 202. If only the headers are expected then it would need a api version upgrade else cause a breaking change. 
Based on the earlier response, we'll remove 202 and go with 201 for future api versions and take this as an action item.

**Comment by meet1993shah** [comment]
> > ```
> > "CloudError": {
> > ```
> 
> I recommend referencing the common type json.
> 
> Refers to: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:3252 in [beef77b](https://github.com/Azure/azure-rest-api-specs/commit/beef77bc1167e286f9f104d85e4a97388dda9e70). [](commit_id = [beef77b](https://github.com/Azure/azure-rest-api-specs/commit/beef77bc1167e286f9f104d85e4a97388dda9e70), deletion_comment = False)

Our definition of CloudError is slightly different then the common-type as we have innerError in the schema. We'll take this is an action item for the future api version.

**Comment by meet1993shah** [comment]
> > ```
> > },
> > ```
> 
> I would recommend referencing the common type json.
> 
> Refers to: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/disk.json:1728 in [beef77b](https://github.com/Azure/azure-rest-api-specs/commit/beef77bc1167e286f9f104d85e4a97388dda9e70). [](commit_id = [beef77b](https://github.com/Azure/azure-rest-api-specs/commit/beef77bc1167e286f9f104d85e4a97388dda9e70), deletion_comment = False)

This is an existing schema and not a new change. Arm and disk both use this in response so changing this base reference causes a breaking change.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateADiskEncryptionSet.json (hunk 8 above)
Comment (j5lim) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateADiskEncryptionSet.json:25]: It won't require an api-version change since the schema for this version already defines Type and Id.

**Review by j5lim — APPROVED** [review]
Approved from ARM side. Please follow the current RPC in a future api version.

File: specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateADiskEncryptionSet.json (hunk 8 above)
Comment (chetmaddula) [reply specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/examples/CreateADiskEncryptionSet.json:25]: the diskrp server side code doesn't return it currently.. so this would need a code change and deployment
